### DISCUSSION TITLE
Implementation of deployment module

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2722] [Q&A] [answered]
I need help using the `deployment` module.

I added `MODULES += $(CONTIKI_NG_SERVICES_DIR)/deployment` to my project `Makefile`

Added `#define DEPLOYMENT_MAPPING deployment_cooja8` to my `project-conf.h`

In one of my source files (I choose the node that is my RPL Root node because only one of it is instantiated)
1. Included `deployment.h`
2. Copied the mapping from `examples/libs/deployment/node.c` :
````
include "deployment.h"

const struct id_mac deployment_cooja8[] = {
  {  1, {{0x00,0x01,0x00,0x01,0x00,0x01,0x00,0x01}}},
  {  2, {{0x00,0x02,0x00,0x02,0x00,0x02,0x00,0x02}}},
  {  3, {{0x00,0x03,0x00,0x03,0x00,0x03,0x00,0x03}}},
  {  4, {{0x00,0x04,0x00,0x04,0x00,0x04,0x00,0x04}}},
  {  5, {{0x00,0x05,0x00,0x05,0x00,0x05,0x00,0x05}}},
  {  6, {{0x00,0x06,0x00,0x06,0x00,0x06,0x00,0x06}}},
  {  7, {{0x00,0x07,0x00,0x07,0x00,0x07,0x00,0x07}}},
  {  8, {{0x00,0x08,0x00,0x08,0x00,0x08,0x00,0x08}}},
  {  0, {{0}}}
};
````

After `make distclean`, compiling for `.native` yields:
````
make TARGET=native
...
...

  LD        build/native/rpl-root.native
  CP        build/native/rpl-root.native --> rpl-root.native
  CC        controller.c
  LD        build/native/controller.native
/usr/bin/ld: build/native/obj/deployment.o: warning: relocation against `deployment_cooja8' in read-only section `.text'
/usr/bin/ld: build/native/obj/deployment.o: in function `deployment_init':
deployment.c:(.text+0xf): undefined reference to `deployment_cooja8'
/usr/bin/ld: build/native/obj/deployment.o: in function `deployment_id_from_lladdr':
deployment.c:(.text+0x80): undefined reference to `deployment_cooja8'
/usr/bin/ld: build/native/obj/deployment.o: in function `deployment_lladdr_from_id':
deployment.c:(.text+0xeb): undefined reference to `deployment_cooja8'
/usr/bin/ld: build/native/obj/deployment.o: in function `deployment_id_from_index':
deployment.c:(.text+0x22e): undefined reference to `deployment_cooja8'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
make: *** [../../Makefile.include:582: build/native/controller.native] Error 1
````

**Comment by rahav** [comment]
I found a solution and I think that I know what is going on. Every source code that appears in `CONTIKI_PROJECT` in your project `Makefile` must `#include "deployment.h"` and must repeat the definition of the deployment mapping table.

Reading through Contiki node instantiation code, I see the reason for this. Every node (equivalently every source code in the project) must use the mapping table to instantiate. I don't know if there is another solution, I'll think about it, but this does not seem optimal because the table will be part of all executables.

**Reply by pjonsson** [reply, marked as answer]
There shouldn't be a need to repeat the table, put the table in one file, and make sure that file is linked into all binaries.

Unless you are running a massive deployment, or deploying on a tiny platform (Sky/Z1, not a 32-bit platform), the size of the table should be negligible.
